### PR TITLE
[Fabric-Sync] Port 'add-bridge' and 'remove-bridge' commands

### PR DESCRIPTION
**Pull request by yufengwangca**
Port 'add-bridge' and 'remove-bridge' commands from fabric-admin to fabric-sync app.

### CODE REVIEW

File: examples/fabric-sync/main.cpp [hunk 1]
```diff
@@ -35,6 +40,21 @@ int main(int argc, char * argv[])
 
     VerifyOrDie(ChipLinuxAppInit(argc, argv) == 0);
 
+#if defined(ENABLE_CHIP_SHELL)
+#if CHIP_DEVICE_CONFIG_APP_PLATFORM_ENABLED
```
Comment (tehampson) [inline examples/fabric-sync/main.cpp:44]: Not sure I follow why there is a check for `CHIP_DEVICE_CONFIG_APP_PLATFORM_ENABLED`

File: examples/fabric-sync/shell/ShellCommands.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+#include <lib/support/CHIPArgParser.hpp>
+#include <lib/support/CHIPMem.h>
+#include <lib/support/CodeUtils.h>
+#include <platform/CHIPDeviceLayer.h>
+
+namespace chip {
+namespace Shell {
+
+// Global pointer to hold the AddBridgeCommand instance
+static std::unique_ptr<AddBridgeCommand> gAddBridgeCommand;
```
Comment (tehampson) [inline examples/fabric-sync/shell/ShellCommands.cpp:37]: What is this for?

File: examples/fabric-sync/admin/DeviceManager.h [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ */
+
+#pragma once
+
+#include "PairingManager.h"
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <platform/CHIPDeviceLayer.h>
+
+class DeviceManager : public PairingDelegate
```
Comment (tehampson) [inline examples/fabric-sync/admin/DeviceManager.h:26]: I always thought that this class could benefit with documentation to let reader know this manages devices on the remote bridge and not on the local bridge.

Do you think it is worth adding now?

File: examples/fabric-sync/admin/DeviceManager.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        return err;
+    }
+
+    ChipLogProgress(NotSpecified, "Successfully paired remote fabric bridge: Node ID " ChipLogFormatX64, ChipLogValueX64(nodeId));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeviceManager::UnpairRemoteFabricBridge()
+{
+    CHIP_ERROR err = PairingManager::Instance().UnpairDevice(mRemoteBridgeNodeId);
```
Comment (tehampson) [inline examples/fabric-sync/admin/DeviceManager.cpp:83]: Should we check that mRemoteBridgeNodeId != kUndefinedNodeId?

File: examples/fabric-sync/admin/DeviceManager.cpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(NotSpecified, "Failed to unpair remote bridge device " ChipLogFormatX64, ChipLogValueX64(mRemoteBridgeNodeId));
+        return err;
+    }
+
+    ChipLogProgress(NotSpecified, "Successfully unpaired remote fabric bridge: Node ID " ChipLogFormatX64,
```
Comment (tehampson) [inline examples/fabric-sync/admin/DeviceManager.cpp:90]: If successful, should we set the value of mRemoteBridgeNodeId to kUndefinedNodeId. Or is that expected to be done by calling UnpairDevice?

File: examples/fabric-sync/shell/AddBridgeCommand.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+CHIP_ERROR AddBridgeCommand::RunCommand()
+{
+    if (DeviceMgr().IsFabricSyncReady())
+    {
+        // print to console
+        Shell::streamer_t * sout = Shell::streamer_get();
+        Shell::streamer_printf(sout, "Remote Fabric Bridge has already been configured.\n");
+
+        return CHIP_NO_ERROR;
```
Comment (tehampson) [inline examples/fabric-sync/shell/AddBridgeCommand.cpp:76]: It is not clear to me why this isn't an error like BUSY or something. Someone might be calling add bridge with different credentials and this success might give the wrong impression that it was successful

File: examples/fabric-sync/shell/AddBridgeCommand.cpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+#include "AddBridgeCommand.h"
+
+#include <admin/DeviceManager.h>
+#include <lib/shell/streamer.h>
+
+using namespace ::chip;
+
+// Define the global unique pointer
+std::unique_ptr<AddBridgeCommand> gAddBridgeCmd = nullptr;
```
Comment (tehampson) [inline examples/fabric-sync/shell/AddBridgeCommand.cpp:27]: Is there any way we can avoid a global here? If not can we added a comment on why things need to be this way?

File: examples/fabric-sync/shell/RemoveBridgeCommand.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+#include "RemoveBridgeCommand.h"
+
+#include <admin/DeviceManager.h>
+#include <lib/shell/streamer.h>
+
+using namespace ::chip;
+
+// Define the global unique pointer
+std::unique_ptr<RemoveBridgeCommand> gRemoveBridgeCmd = nullptr;
```
Comment (tehampson) [inline examples/fabric-sync/shell/RemoveBridgeCommand.cpp:27]: Same, Is there any way we can avoid a global here? If not, can we added a comment on why things need to be this way?

File: examples/fabric-sync/shell/ShellCommands.cpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+static CHIP_ERROR HandleRemoveBridgeCommand(int argc, char ** argv)
+{
+    if (argc != 1)
+    {
+        fprintf(stderr, "Invalid arguments. Usage: app remove-bridge\n");
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    gRemoveBridgeCmd.reset();
```
Comment (tehampson) [inline examples/fabric-sync/shell/ShellCommands.cpp:82]: Is this safe? What happens if you reset here, but the RunCommand below fails and doesn't end up calling SetPairingDelegate. Would you have a use after free?

Essentially should there be a dtor for RemoveBridgeCommand

File: examples/fabric-sync/shell/ShellCommands.cpp [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    {
+        return HandleAddBridgeCommand(argc, argv);
+    }
+    else if (strcmp(argv[0], "remove-bridge") == 0)
+    {
+        return HandleRemoveBridgeCommand(argc, argv);
+    }
+    else if (strcmp(argv[0], "sync-device") == 0)
+    {
+        // TODO
```
Comment (tehampson) [inline examples/fabric-sync/shell/ShellCommands.cpp:110]: empty TODO

File: examples/fabric-sync/main.cpp (hunk 1 above)
Comment (yufengwangca) [reply examples/fabric-sync/main.cpp:44]: The Shell commission commands depend on examples/platform/linux/CommissionerMain.cpp to implement their functionalities, which require this flag. Without it, the following compilation errors will occur:

```
../../../examples/fabric-sync/third_party/connectedhomeip/examples/platform/linux/CommissionerMain.cpp: In constructor ‘PairingCommand::PairingCommand()’:
../../../examples/fabric-sync/third_party/connectedhomeip/examples/platform/linux/CommissionerMain.cpp:243:9: error: class ‘PairingCommand’ does not have any field named ‘mOnDeviceConnectedCallback’
  243 |         mOnDeviceConnectedCallback(OnDeviceConnectedFn, this),
```

Comment (tehampson) [reply examples/fabric-sync/main.cpp:44]: Would some sort of static assert here be better? Is the app going to have some functionality without being able to call `Shell::RegisterCommands();`?

Essentially if I enable `ENABLE_CHIP_SHELL`, but don't enable `CHIP_DEVICE_CONFIG_APP_PLATFORM_ENABLED`, why do we let things still compile that prevents `ENABLE_CHIP_SHELL` from working?

File: examples/fabric-sync/admin/DeviceManager.cpp (hunk 5 above)
Comment (yufengwangca) [reply examples/fabric-sync/admin/DeviceManager.cpp:90]: This is done in void RemoveBridgeCommand::OnDeviceRemoved(NodeId deviceId, CHIP_ERROR err)

File: examples/fabric-sync/admin/DeviceManager.cpp (hunk 4 above)
Comment (yufengwangca) [reply examples/fabric-sync/admin/DeviceManager.cpp:83]: acked

File: examples/fabric-sync/shell/ShellCommands.cpp (hunk 2 above)
Comment (yufengwangca) [reply examples/fabric-sync/shell/ShellCommands.cpp:37]: removed

File: examples/fabric-sync/admin/DeviceManager.h (hunk 3 above)
Comment (yufengwangca) [reply examples/fabric-sync/admin/DeviceManager.h:26]: In fabric-admin, DeviceManager manages both local bridge and remote bridge, in fabric-sync, it currently only manages the remote bridge for 1-1 fabric-sync, but we might also need to manage the local bridge later, lets keep it as it is now.

**Review by tehampson — APPROVED** [review]

**Review by joonhaengHeo — APPROVED** [review]

**Pull request merged**